### PR TITLE
feat: deploy with config on local cluster

### PR DESCRIPTION
**Pull request by ericyangchen**
feat: deploy idas service on local cluster #1

### CODE REVIEW

File: handler/deployment_config_callback.go [hunk 1]
```diff
@@ -181,14 +186,175 @@ func (h *handler) ReceiveDeleteDeploymentConfig(s *discordgo.Session, i *discord
...
+		}); err != nil {
+			log.Println(err)
+		}
+		return
+	}
+
+	// Create the Signer for this private key.
+	signer, err := ssh.ParsePrivateKey(pKey)
+	if err != nil {
+		fmt.Println(err.Error())
```
Comment (DenChenn) [inline handler/deployment_config_callback.go:217]: after log out the error, u should `return`

File: handler/deployment_config_callback.go [hunk 2]
```diff
@@ -181,14 +186,175 @@ func (h *handler) ReceiveDeleteDeploymentConfig(s *discordgo.Session, i *discord
...
+			},
+		}); err != nil {
+			log.Println(err)
+		}
+		return
+	}
+
+	// Connect to the remote server
+	err = scpClient.Connect()
+	if err != nil {
```
Comment (DenChenn) [inline handler/deployment_config_callback.go]:
```suggestion
	if err := scpClient.Connect(); err != nil {
```

File: handler/deployment_config_callback.go [hunk 3]
```diff
@@ -181,14 +186,175 @@ func (h *handler) ReceiveDeleteDeploymentConfig(s *discordgo.Session, i *discord
...
+			Type: discordgo.InteractionResponseChannelMessageWithSource,
+			Data: &discordgo.InteractionResponseData{
+				Content: "❌ Failed to connect to server over ssh",
+			},
+		}); err != nil {
+			log.Println(err)
+		}
+		return
+	}
+	// Close client connection after the file has been copied
```
Comment (DenChenn) [inline handler/deployment_config_callback.go]: This annotation is wrong 🥲
The client will close after the function finished.

File: handler/deployment_config_callback.go [hunk 4]
```diff
@@ -181,14 +186,175 @@ func (h *handler) ReceiveDeleteDeploymentConfig(s *discordgo.Session, i *discord
...
+	}
+	file, err := os.Open(yamlFilePath)
+	if err != nil {
+		log.Println(err)
+	}
+	filename := deploymentConfig.Name + "_" + deploymentConfig.ID + ".yaml"
+	remoteFilePath := filepath.Join(serverDeploymentFileRoot, filename)
+	permission := "0655"
+	err = scpClient.CopyFromFile(context.Background(), *file, remoteFilePath, permission)
+	if err != nil {
```
Comment (DenChenn) [inline handler/deployment_config_callback.go]:
```suggestion
	if err := scpClient.CopyFromFile(context.Background(), *file, remoteFilePath, "0655"); err != nil {
```

File: handler/deployment_config_callback.go [hunk 5]
```diff
@@ -181,14 +186,175 @@ func (h *handler) ReceiveDeleteDeploymentConfig(s *discordgo.Session, i *discord
...
+		log.Println(err)
+		if err := s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
+			Type: discordgo.InteractionResponseChannelMessageWithSource,
+			Data: &discordgo.InteractionResponseData{
+				Content: "❌ Failed to copy file to server",
+			},
+		}); err != nil {
+			log.Println(err)
+		}
+		template.RemoveYaml(yamlFilePath)
```
Comment (DenChenn) [inline handler/deployment_config_callback.go]: You should yaml file remove after file-copy process success.

File: handler/deployment_config_callback.go (hunk 3 above)
Comment (DenChenn) [reply handler/deployment_config_callback.go]: Moreover, try to add `defer` after creating client. (Not after error handling...
Imagine the situation that some error happened above, the client will never be closed...

**Review by DenChenn — CHANGES_REQUESTED** [review]

File: handler/deployment_config_callback.go [hunk 6]
```diff
@@ -181,14 +186,172 @@ func (h *handler) ReceiveDeleteDeploymentConfig(s *discordgo.Session, i *discord
...
+				Content: "❌ Failed to connect to server over ssh",
+			},
+		}); err != nil {
+			log.Println(err)
+		}
+		return
+	}
+
+	// kubectl apply file
+	cmd := "kubectl apply " + serverDeploymentFileRoot
```
Comment (DenChenn) [inline handler/deployment_config_callback.go]: this should be `kubectl apply -f` ?

File: handler/deployment_config_callback.go [hunk 7]
```diff
@@ -181,14 +186,172 @@ func (h *handler) ReceiveDeleteDeploymentConfig(s *discordgo.Session, i *discord
...
+				Content: "❌ Failed to apply deployment config",
+			},
+		}); err != nil {
+			log.Println(err)
+		}
+		template.RemoveYaml(yamlFilePath)
+		return
+	}
+	// TODO: retrieve and send kubectl log to client
+	log.Println("[Kubectl Log]:\n", string(result))
```
Comment (DenChenn) [inline handler/deployment_config_callback.go]: This one shouldn't be logged into backend server... or our log server will be flooded...
Btw, logging is a REALLY EXPENSIVE SERVICE.

Comment (ericyangchen) [reply handler/deployment_config_callback.go]: yeah, it should be removed
I added just for testing purposes

File: handler/deployment_config_callback.go [hunk 8]
```diff
@@ -181,14 +186,172 @@ func (h *handler) ReceiveDeleteDeploymentConfig(s *discordgo.Session, i *discord
...
+		}); err != nil {
+			log.Println(err)
+		}
+		return
+	}
+
+	// kubectl apply file
+	cmd := "kubectl apply -f " + serverDeploymentFileRoot
+	_, err = session.Output(cmd)
+	if err != nil {
```
Comment (DenChenn) [inline handler/deployment_config_callback.go]:
```suggestion
	if _, err := session.Output(cmd); err != nil {
```

File: handler/deployment_config_callback.go [hunk 9]
```diff
@@ -181,14 +186,172 @@ func (h *handler) ReceiveDeleteDeploymentConfig(s *discordgo.Session, i *discord
...
+		log.Println(err)
+		if err := s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
+			Type: discordgo.InteractionResponseChannelMessageWithSource,
+			Data: &discordgo.InteractionResponseData{
+				Content: "❌ Failed to apply deployment config",
+			},
+		}); err != nil {
+			log.Println(err)
+		}
+		template.RemoveYaml(yamlFilePath)
```
Comment (DenChenn) [inline handler/deployment_config_callback.go]: Yaml removal should be done outside of err

File: handler/deployment_config_callback.go (hunk 5 above)
Comment (DenChenn) [reply handler/deployment_config_callback.go]: Not fix

**Review by DenChenn — APPROVED** [review]

**Pull request merged**